polynote-runtime/src/main/scala/polynote/runtime/python/PythonObject.scala

```diff
@@ -0,0 +1,127 @@
+
+import scala.language.dynamics
+
+/**
+  * Just a bit of Scala sugar over [[PyObject]]
+  */
+class PythonObject(obj: PyObject, runner: PythonObject.Runner) extends Dynamic {
+  import PythonObject.unwrapArg
+
+  private[polynote] def unwrap: PyObject = obj
+
+  protected def callPosArgs(callable: PyCallable, args: Seq[AnyRef]): AnyRef = {
+    runner.run {
+      callable.call(args.map(unwrapArg): _*)
+    } match {
+      case pc: PyCallable => new PythonFunction(pc, runner)
+      case po: PyObject   => new PythonObject(po, runner)
+      case other => other
+    }
+  }
+
+  protected def callKwArgs(callable: PyCallable, args: Seq[(String, Any)]): AnyRef = {
+    val kwArgIndex = args.indexWhere(_._1.nonEmpty)
+    val (posArgs, kwArgs) = if (kwArgIndex >= 0) {
+      args.splitAt(kwArgIndex)
+    } else (args, Nil)
+
+    val posArgsArray = posArgs.map(tup => unwrapArg(tup._2.asInstanceOf[AnyRef])).toArray
+    val kwArgsMap = kwArgs.toMap.mapValues(_.asInstanceOf[AnyRef]).mapValues(unwrapArg).toSeq.toMap // make sure it's a strict map
+
+    runner.run {
+      if (posArgsArray.nonEmpty && kwArgsMap.nonEmpty) {
+        callable.call(posArgsArray, kwArgsMap.asJava)
+      } else if (posArgsArray.nonEmpty) {
+        callable.call(posArgsArray: _*)
+      } else if (kwArgsMap.nonEmpty) {
+        callable.call(kwArgsMap.asJava)
+      } else {
+        callable.call()
+      }
+    } match {
+      case pc: PyCallable => new PythonFunction(pc, runner)
+      case po: PyObject   => new PythonObject(po, runner)
+      case other => other
+    }
+  }
+
+  def selectDynamic[T](name: String)(implicit returnType: ReturnTypeFor[T]): returnType.Out = {
+    runner.run {
+      returnType.wrap(obj.getAttr(name, returnType.tag.runtimeClass.asInstanceOf[Class[returnType.Class]]), runner)
+    }
+  }
+
+  def updateDynamic(name: String)(value: Any): Unit = runner.run {
+    obj.setAttr(name, value)
+  }
+
+  // TODO: Jep doesn't give us a way to invoke a PyCallable and get back a PyObject in case the result can't be converted
+  //       to a Java representation. So we'll run into the stringification problem here. We should try to fix this upstream
+  //       in Jep.
+  def applyDynamic(method: String)(args: Any*): Any =
+      callPosArgs(selectDynamic[PyCallable](method), args.asInstanceOf[Seq[AnyRef]])
+
+
+  def applyDynamicNamed(method: String)(args: (String, Any)*): Any =
+      callKwArgs(selectDynamic[PyCallable](method), args)
+
+  override def toString: String = runner.run {
+    obj.toString
+  }
+
+}
+
+object PythonObject {
+
+  def unwrapArg(arg: AnyRef): AnyRef = arg match {
+    case pyObj: PythonObject => pyObj.unwrap
+    case other => other
+  }
+
+  // We can't have an overloaded selectDynamic, so this is an approximation of it.
+  // When called with no type argument, the type argument is inferred to be Nothing, so we can use this type function
+  // to catch that case and return PythonObject.
```

Comment:
LMK if the reason for this typeclass needs further explanation

Reply:
> We can't have an overloaded selectDynamic, so this is an approximation of it.
Since the calls to selectDynamic are filled in by the compiler, we can't mess with the way it looks by adding params, etc. However, since implicit params don't change that, we can add them. I guess this is because the transformation of `foo.bar` to `foo.selectDynamic("bar")` is done before the compiler looks for implicits and thus they'll be properly handled?

^ if that's correct, then the explanation you have is fine :)

Reply:
Yes, the issue is that you can't have both `selectDynamic()` and `selectDynamic[T]()` because the call is ambiguous for the first one. So you have to always use `selectDynamic[T]`, but the problem is if `T` is omitted we want the return type to be `PythonObject`. But your description of the dynamic rewriting is accurate.